pkg/controller/atlasdeployment/searchindex_test.go

```diff
@@ -57,32 +59,25 @@ func Test_searchIndexReconciler(t *testing.T) {
 			Status: status.AtlasDeploymentStatus{},
 		}
 
-		mockSearchAPI := mockadmin.NewAtlasSearchApi(t)
-		mockSearchAPI.EXPECT().
-			CreateAtlasSearchIndex(context.Background(), mock.Anything, mock.Anything, mock.Anything).
-			Return(admin.CreateAtlasSearchIndexApiRequest{ApiService: mockSearchAPI})
-		mockSearchAPI.EXPECT().
-			CreateAtlasSearchIndexExecute(admin.CreateAtlasSearchIndexApiRequest{ApiService: mockSearchAPI}).
-			Return(&admin.ClusterSearchIndex{
-				CollectionName: "",
-				Database:       "",
-				IndexID:        pointer.MakePtr("testID"),
-				Name:           "",
-				Status:         pointer.MakePtr("NOT STARTED"),
-			}, &http.Response{StatusCode: http.StatusCreated}, nil)
+		fakeAtlasSearch := &fake.FakeAtlasSearch{
```

Comment:
Why did you add this custom mock? what's the problem with using the SDK mock?

Reply:
Having the translation layer means we can chose to test more easily at the translation layer interface.

Testing at the API level is not only more verbose and awkward, but also redundant in most cases, because the translation layer logic is covered by contract tests and conversion unit tests.

Still, we might want to use the SDK mocks at times. For instance, this migration translated tests that were validating logic that was moved within the translation layer, and for those tests the SDK mocks needed to be used as the test had to reach inside said translation layer. For the rest of the test that just assume a certain reply by the API but which validate code at the reconciliation logic, the simpler translation layer interface is preferred.

Reply:
The SDK mocks still have a lot to improve, but I'm afraid that running away from it on our unit tests might lead to confusion on how to mock API calls.

Reply:
There should not be confusion, we now have a layer between the reconciliation logic and the API calls. Unit tests should not try to test 2 layers at once when they can avoid it, or unless there is a very good reason for it:

The layers we have are:

1. **Reconcile logic** - Unit tests itself against the translation layer interface, preferably. Probably always on new code.
2. **Translation layer** - Unit tests itself (when needed) against the SDK mocks. Still most conversion unit tests may not need to mock the SDK and Contract Tests do not mock the SDK at all. But if we need to unit test a full API call for whatever reason, like some tricky logic within that is not just conversions, the SDK mocks are the ones to use here.
3. **SDK**

Reply:
The proper way would be to mock the translation layer service, but you are here adding a custom mock that does the same as the SDK mock. I see no reason to replace the SDK mock if you're still faking the internal calls abstracted by the translation layer. IMO, you either mock translation service or use SDK mocks

Reply:
This new fake is for the translation layer service interface:

```go
type FakeAtlasSearch struct {
	GetIndexFn    func(ctx context.Context, projectID, clusterName, indexName, indexID string) (*searchindex.SearchIndex, error)
	CreateIndexFn func(ctx context.Context, projectID, clusterName string, index *searchindex.SearchIndex) (*searchindex.SearchIndex, error)
	DeleteIndexFn func(ctx context.Context, projectID, clusterName, indexID string) error
	UpdateIndexFn func(ctx context.Context, projectID, clusterName string, index *searchindex.SearchIndex) (*searchindex.SearchIndex, error)
}
```

This is NOT the same as the Atlas SDK API call.

For instance, here we are mocking:
```go
CreateIndex(context.Context, string, string, *searchindex.SearchIndex) (*searchindex.SearchIndex, error)
```
Instead of:
```go
CreateAtlasSearchIndex(context.Context, string, string, *admin.ClusterSearchIndex) (admin.CreateAtlasSearchIndexApiRequest)
+
CreateAtlasSearchIndexExecute(admin.CreateAtlasSearchIndexApiRequest)
```

It is not the same. It is a thin layer, but it is a layer.